src/hooks/fire/useLiveCollection.ts

```diff
@@ -0,0 +1,103 @@
+import { useEffect, useMemo, useState } from "react";
+import { QueryConstraint } from "@firebase/firestore";
+import {
+  collection,
+  FirestoreError,
+  getFirestore,
+  onSnapshot,
+  query,
+  QueryDocumentSnapshot,
+  QuerySnapshot,
+} from "firebase/firestore";
+import { isEqual } from "lodash";
+
+import { ALWAYS_EMPTY_ARRAY } from "settings";
+
+import { withIdConverter } from "utils/converters";
+import { withId } from "utils/id";
+
+type UseLiveCollectionOptions =
+  | string[]
+  | {
+      path: string[];
+      constraints?: (QueryConstraint | null | undefined)[];
+    };
+
+const toDataWithId = <T extends object>(d: QueryDocumentSnapshot<T>) =>
+  withId(d.data(), d.id);
+
+const isValidSegment = (segment: string | undefined): segment is string =>
+  "string" === typeof segment && "" !== segment;
+
+const isValidConstraint = (
+  constraint: QueryConstraint | null | undefined
+): constraint is QueryConstraint =>
+  constraint !== null && constraint !== undefined;
+
+export const useLiveCollection = <T extends object>(
+  options: UseLiveCollectionOptions
+) => {
+  const [data, setData] = useState<T[]>();
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<Error | undefined>();
+
+  const path = Array.isArray(options) ? options : options?.path;
+  const constraints = Array.isArray(options) ? undefined : options?.constraints;
+
+  const memoizedPath = useMemo(
+    () => (path ?? ALWAYS_EMPTY_ARRAY).filter(isValidSegment),
+    [path]
+  );
+
+  const memoizedConstraints = useMemo(
+    () => (constraints ?? ALWAYS_EMPTY_ARRAY).filter(isValidConstraint),
+    [constraints]
+  );
```

Comment:
I think quietly allowing for bad queries by filtering out broken parts is going to leave to hard to find bugs. If I write a query that is wrong I expect to have a very noisy error as soon as I try it.

Reply:
Similarly for the other new hooks.

Reply:
>  If I write a query that is wrong I expect to have a very noisy error as soon as I try it.

And if I write a query that has missing `spaceSlug` because the previous hook is still loading I expect some noisy error to not break the entire page.

These expectations, however, aren't objective, but subjective.

What objectively can be done is to define what is "a query that is wrong" in the first place. 

Then, following from there, add proper tools, like switching the moniker from `undefined`, to some `Symbol()` or some other method of dealing with expected vs unexpected definition of a query, like populate the `error` field that every result has with a `SparkleError` type etc.

Reply:
If we're still waiting for a previous hook then we should carry on waiting rather than doing queries with crap arguments

Reply:
Well, it's not like waiting, it's more like "bad connection". 

From 945342eb8af0a1591a833f0c785c70b1c61b60aa there is now a value named `DEFERRED` to signal the hook should wait for data before trying to connect. The simplest use would be 
```tsx
useLiveDocument(DEFERRED);
```
but usually some `if-else` check that either produces the correct path or puts DEFERRED in one of several possible places.